cuda_bindings/cuda/bindings/_bindings/cydriver.pyx.in

```diff
@@ -486,16 +486,19 @@ cdef bint __cuPythonInit = False
 {{if True}}cdef void *__cuGraphicsVDPAURegisterVideoSurface = NULL{{endif}}
 {{if True}}cdef void *__cuGraphicsVDPAURegisterOutputSurface = NULL{{endif}}
 
+# To make cuPythonInit reentrant
+ctypedef CUresult (*__cuGetProcAddress_v2_T)(const char*, void**, int, cuuint64_t, CUdriverProcAddressQueryResult*) except?CUDA_ERROR_NOT_FOUND nogil
+cdef __cuGetProcAddress_v2_T _F_cuGetProcAddress_v2 = NULL
+
 cdef int cuPythonInit() except -1 nogil:
     global __cuPythonInit
+    if __cuPythonInit:
+        return 0
```

Comment:
With this outside of the lock if you have a bunch of threads trying to run this initially at the same time, you can end up in the situation where they all don't hit this early exit and all wait for the symbol lock and the reinitialize all of the symbols. I had done a quick and dirty local benchmark with a single thread acquiring a lock and doing nothing to understand the overhead in the early exit case and it was ~50ns.

Reply:
Yeah I thought about this too. Good to know it's only ~50 ns. We should fix the codegen so that we check if symbol is null to avoid re-initialization, but it can be done later.